### ISSUE TITLE
Add S3 Bucket to SNS mapping

### ISSUE THREAD

**Issue by AlexeyRaga** [hashicorp/terraform#4120]
It would be very helpful if Terraform could add event notifications from an S3 Bucket to an SNS topic.
ASW has this ability through the UI and it is also supported by CloudFormation, but not by Terraform as far as I can tell.

**Comment by stack72** [comment]
@AlexeyRaga I cannot see, within the UI, a way to hook an S3 bucket up to an SNS Topic

<img width="684" alt="screen shot 2015-12-01 at 13 21 53" src="https://cloud.githubusercontent.com/assets/227823/11501746/904b91e4-982e-11e5-9100-0bab73c4669e.png">

Got any more info on how to do this?

**Comment by wstaples** [comment]
From the S3 console click properties and expend the Events section:


![untitled](https://cloud.githubusercontent.com/assets/3649913/11568744/ec738f06-99bb-11e5-9c46-3a8267a6ea08.png)



You can also use the cli:
`aws s3api get-bucket-notification-configuration --bucket <some_bucket>`

The best way to handle this would probably be to add an event property to the aws_s3_bucket resource and users can provide there own JSON.

**Comment by dwradcliffe** [comment]
Yeah bucket event config would be a big help!

**Comment by dwradcliffe** [comment]
See also: https://github.com/hashicorp/terraform/issues/3108

**Comment by mdevs5531** [comment]
+1

**Comment by comozo** [comment]
+1 for s3 events -> SNS

**Comment by harrisnb** [comment]
+1 for s3 events -> SNS allowing users to provide their own JSON

**Comment by dabdine-r7** [comment]
I was able to work around this by combining an idea from @wstaples and the ```null_resource```, a trigger and the ```local_exec``` provisioner:

```
resource "null_resource" "test_bucket_sns_event_notification" {
   triggers {
      bucket = "${aws_s3_bucket.test_bucket.arn}"
   }

   provisioner "local-exec" {
      command = <<EOF
aws s3api put-bucket-notification-configuration --region ${var.region} --bucket ${aws_s3_bucket.test_bucket.bucket} --notification-configuration \
'{
    "TopicConfigurations": [
        {
            "TopicArn": "${aws_sns_topic.some_sns_topic.arn}",
            "Events": [
                "s3:ObjectCreated:*"
            ]
        }
    ]
}'
EOF
   }
}
```

Obvious pros and cons here, so YMMV.

CONS:
1. Terraform can't enforce policy configuration in this model so be aware of users hand-modifying the policy via the aws cli OR the aws console.
2. Requires the aws cli to be installed locally to make use of s3api.

PROS: 
It works :P

**Comment by dabdine-r7** [comment]
Oh, and +1 for s3 events -> SNS

**Comment by sp-ludovic-ivain** [comment]
+1 for s3 events -> SQS :)

**Comment by joseph-wortmann** [comment]
+1 for s3 events -> SQS

**Comment by gilessteinberg** [comment]
+1 for s3 events -> Lambda

**Comment by wincus** [comment]
+1 for s3 events -> SQS

**Comment by falaa** [comment]
+1 for s3 events -> SQS

**Comment by egoldschmidt** [comment]
+1

**Comment by iceycake** [comment]
S3 Event Notification will look something like this?

```
resource "aws_s3_bucket" "b" {
  notification {
    lambda {
        event = ""
        function = ""
        filter {
            name = ""
            value = ""
        }
        filter {
            name = ""
            value = ""
        }
    }
    sns {
        event = ""
        topic = ""
        filter {
          name = ""
          value = ""
        }
    }
    sqs {
        event = ""
        queue = ""
        filter {
          name = ""
          value = ""
        }
    }
  }
}
```

**Comment by conorgil** [comment]
+1 for s3 events -> SQS

**Comment by rantav** [comment]
Joining the herd: 
+1 for s3 events -> Lambda

**Comment by rbachman** [comment]
+1 for s3 events

**Comment by curena** [comment]
It looks like bucket event notifications are being implemented via https://github.com/hashicorp/terraform/pull/5473

**Comment by stack72** [comment]
This has been implemented via #5473 with thanks to @kjmkznr
